### ISSUE TITLE
Bugs 

### ISSUE THREAD

**Issue by seal** [Root-Branch/cardamon#38]
This issue will contain current bugs I have found:
## Need cancellation in keep_logging 
in src/metrics_logger/docker.rs:28
```rust
pub async fn keep_logging(container_names: Vec<String>, metrics_log: Arc<Mutex<MetricsLog>>) {
```
This function is meant to flush the buffer every *x* times *or* if the function is cancelled.
Currently it flushes every 5 "logs", but if there are *3* logs and the function is cancelled ( scenario *finished* )  no logs will be flushed ( this is an issue because cadvisor updates every ( approx ) 2 seconds, meaning any scenario under 10s will have *no* logs ) 

### Suggested fix:
code change in src/metric_logger.rs:
```rust
 join_set.spawn(async move {
            tokio::select! {
                _ = token.cancelled() => {}
                _ = docker::keep_logging(
                        container_names,
                        shared_metrics_log,
                    ) => {}
            }
        });

```
This currently has the tokio:select inside the start_logging function, I'd suggest passing the token as cloned value into keep_logging
e.g. 
```rust
 join_set
            .spawn(async move { _ = docker::keep_logging(token,container_names, shared_metrics_log) });
```

This means I can do: ( pseudo-ish-code) 
```rust
        iteration_count += 1;

        if iteration_count == 5 || token.cancelled() {
            let mut metrics_log = metrics_log
                .lock()
                .expect("Failed to acquire lock on MetricsLog");
            for metric in buffer.drain(..) {
                metrics_log.push_metrics(metric);
            }
            for error in error_buffer.drain(..) {
                metrics_log.push_error(error);
            }
            debug!("Flushed metrics and errors to MetricsLog");

            iteration_count = 0;

            if token.cancelled() {
                info!("Token cancelled. Exiting keep_logging loop.");
                break;
            }
        }
```

 
## Cpu Metrics 
cpu metrics, found in src/metrics.rs:
```rust
#[derive(Debug, Clone)]
pub struct CpuMetrics {
    pub process_id: String,
    pub process_name: String,
    pub cpu_usage: f64,
    pub core_count: i32,
    pub timestamp: i64,
}
```
Currently the *docker* metrics gathering uses this type, the only issue with this is "process_id" and "process_name" don't apply to containers, I'd suggest using a more general name "id" and "name" 

## clap.rs
Current usage of the program is :
```
cargo run --bin card -- {observation_name}
```
I believe this naming convention is confusing and should be replace with 
--name={observation_name} 

## DB Not found 
db connection, found in src/main.rs:39
```rust
 let pool = sqlx::sqlite::SqlitePoolOptions::new()
                .max_connections(4)
                .connect("sqlite://cardamon.db")
                .await?;
```
This code *fails* if there is no cardamon.db found in the file system, I'd suggest creating the database if there's an error.
e.g. 
```rust
 if !sqlx::Sqlite::database_exists("sqlite://cardamon.db").await? {
                sqlx::Sqlite::create_database("sqlite://cardamon.db").await?;
            }
```

Furthermore, if we create the DB via:
```
touch cardamon.db
```
When running we get the error:
```
Error: Error inserting scenario into db.

Caused by:
    0: error returned from database: (code: 1) no such table: scenario_run
    1: (code: 1) no such table: scenario_run

```
This can be fixed by something like:
https://github.com/launchbadge/sqlx/issues/1114#issuecomment-801151976
( Using cargo_manifest_dir for the migrations folder, otherwise it'll fail, as the CLI won't always be in our source code dir ) 
```
CARGO_MANIFEST_DIR — The directory containing the manifest for the package being built (the package containing the build script). Also note that this is the value of the current working directory of the build script when it starts.
```
Alternatively it can be fixed by running "sqlx migrate run" *but* this only works if the .env contains "DATABASE_URL=sqlite://cardamon.db" 
This isn't a great solution as if we are abstracting away from the user choosing their local database name, we shouldn't have them set it to our specific one in a .env 

### Options aren't optional 
- Current cardamon.toml says:
```
metrics_server_url = "http://cardamon.rootandbranch.io" # Optional - assumes local db if not specifed
```
These values aren't *actually* optional, and will cause program to crash if they are not present 
#### Example fix:
src/config.rs:13-17:
```rust
#[derive(Debug, Deserialize)]
pub struct Config {
    pub debug_level: Option<String>,
    pub metrics_server_url: Option<String>,
    pub processes: Vec<Process>,
    pub scenarios: Vec<Scenario>,
    pub observations: Vec<Observation>,
}
// line 147
#[derive(Debug, Deserialize, PartialEq)]
pub struct Scenario {
    pub name: String,
    pub desc: Option<String>,
    pub command: String,
}

```
These values have been set to Option<..> in the example ( they are currently not, hence the crash ) 
Same as the "iterations" in Observation, it needs to be something like  .unwrap_or_default(1)  ( and be an option ) 

### get_observation({name}) 
src/config.rs:31
```rust
            .context("Observation with name {name} does not exist.")
```
.context doesn't implement the {} formatting, so if the name of an observation is *invalid* it will print:
```
            .context("Observation with name {name} does not exist.")
```
Every time ( also name variable doesn't exist, but observation_name does ) 

#### Example fix:
```rust
 .context(format!(
                "Observation with name {} does not exist.",
                observation_name
            ))
```

### Empty context 
in src/config.rs:89
```rust
 .find(|scenario| &scenario.name == scenario_name)
                .context("")?;
```
Context is empty, if scenarios are named incorrectly no context is given 
#### Example fix:
```rust
                .context(format!("Scenario name {scenario_name} not found"))?;
```

**Comment by ohuu** [comment]
I've fixed the Optional config problem.

**Comment by ohuu** [comment]
cancel token is no longer required. It's handled in the StopHandle

fixed in #30

**Comment by ohuu** [comment]
clap is not an issue because we only need the empty -- when running through cargo
